Content.Shared/Weapons/Ranged/Components/AlternativeFireModesComponent.cs

```diff
@@ -0,0 +1,43 @@
+using Robust.Shared.GameStates;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization;
+using Robust.Shared.Serialization.TypeSerializers.Implementations.Custom.Prototype;
+using System.Linq;
+
+namespace Content.Shared.Weapons.Ranged.Components;
+
+[RegisterComponent, NetworkedComponent]
+[AutoGenerateComponentState]
+public sealed partial class AlternativeFireModesComponent : Component
+{
+    [DataField("fireModes")]
+    [AutoNetworkedField]
+    public List<FireMode> FireModes = new();
+
+    [AutoNetworkedField]
+    public int CurrentFireModeIndex = 0;
+
+    public FireMode? CurrentFireMode
+    {
+        get
+        {
+            if (FireModes == null || !FireModes.Any())
+                return null;
+
+            return FireModes.ElementAtOrDefault(CurrentFireModeIndex);
+        }
+    }
```

Comment:
just store the current fire mode instead of the index. Field also needs commenting and a datafield attribute so it's serialized.